### PR TITLE
feat(cli): add `--only` and `--skip` options

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

This PR renames `--rule` into `--only`  and adds a new option `--skip`.
These two options can be repeated and used together.
When used together, `--skip` takes precedence over `--only` (like `--ignore` takes precedence over `include`).

For example, the following command executes only the `suspicious/noDebugger` rule and all the rules of the `style` group, but skips the `style/useNamingConvention` rule.

```shell
biome lint --only=suspicious/noDebugger --only=style --skip=style/useNamingConvention
```

Using `skip` on a group, skip all rules in the group.
Conversely, using `--only` on a group, enable all rules of the group - even the ones explicitly disabled in the configuration.

## Limitations

Using `--only` disables the diagnostics about unused suppression comments.
Ideally, we should also filter these diagnostics according to the enabled rules.
However, I don't know how to filter them based on that criteria.

## Test Plan

I added tests and updated previous tests.

### CODE REVIEW

**Comment by ematipico** [comment]
>  When used together, --skip overrides --only.

I gave my explanation on Discord, I will copy it too here:

> Regarding the precedence, I think we should be consistent with what we have already. --only and --skip can be seen as the pair ignore/include, where skip -> ignore and only -> include . We document that ignore takes precedence over include, so we should do the same here.

File: crates/biome_service/src/file_handlers/javascript.rs [hunk 1]
```diff
@@ -422,6 +397,58 @@ pub(crate) fn lint(params: LintParams) -> LintResults {
...
+                            .as_ref()
+                            .map(|rules| rules.as_enabled_rules())
+                            .unwrap_or_default()
+                            .into_iter()
+                            .filter(|rule_filter| {
+                                (params.only.is_empty()
+                                    || only_groups.contains(&rule_filter.group()))
+                                    && !skipped_groups.contains(&rule_filter.group())
+                                    && !skipped_rules.contains(rule_filter)
+                            }),
```
Comment (ematipico) [inline crates/biome_service/src/file_handlers/javascript.rs]: Since we're using a `HashSet`, why don't we use methods such as `intersection` and `difference`? That's one of the reasons why we picked this data structure.

Also, why don't we compute these rules at `AnalysisFilter` level? The `AnalysisFilter` is already ready for this scenario, in fact it accepts `disabled_rules`

https://github.com/biomejs/biome/blob/a51ef9d771c526c5df42da0167a65813b7e80854/crates/biome_analyze/src/lib.rs#L822-L832

File: crates/biome_service/src/file_handlers/json.rs [hunk 2]
```diff
@@ -317,38 +318,66 @@ fn lint(params: LintParams) -> LintResults {
...
+                            .unwrap_or_default()
+                            .into_iter()
+                            .filter(|rule_filter| {
+                                (params.only.is_empty()
+                                    || only_groups.contains(&rule_filter.group()))
+                                    && !skipped_groups.contains(&rule_filter.group())
+                                    && !skipped_rules.contains(rule_filter)
+                            }),
+                    );
+                }
```
Comment (ematipico) [inline crates/biome_service/src/file_handlers/json.rs]: This code seems repeated. The same is in `file_handlers/mod.rs`

File: xtask/codegen/src/generate_configuration.rs [hunk 3]
```diff
@@ -258,7 +258,7 @@ pub(crate) fn generate_rules_configuration(mode: Mode) -> Result<()> {
             }
 
             /// Set the severity of the rule to its default.
-            pub fn set_default_severity(&mut self, group: RuleGroup, rule_name: &str) {
+            pub fn set_default_severity_if_off(&mut self, group: RuleGroup, rule_name: &str) {
```
Comment (ematipico) [inline xtask/codegen/src/generate_configuration.rs:261]: We should update the documentation of this function, its name is different, so maybe its semantics changed?

File: crates/biome_service/src/file_handlers/json.rs (hunk 2 above)
Comment (Conaclos) [reply crates/biome_service/src/file_handlers/json.rs]: I think you commented a previous version of the code.

File: crates/biome_cli/src/commands/mod.rs [hunk 4]
```diff
@@ -201,11 +201,21 @@ pub enum BiomeCommand {
...
+        #[bpaf(long("only"), argument("GROUP|RULE"))]
+        only: Vec<RuleSelector>,
+
+        /// Skip the given rule or rule group.
+        ///
+        /// The option overrides the Biome configuration file as follows:
+        ///
+        /// - When a rule is passed, its severity level is set to `off'.
+        ///
+        /// - When a rule group is passed, all rules of the group are skipped.
```
[the remark this answers is not in the document]
Comment (Conaclos) [reply crates/biome_cli/src/commands/mod.rs:214]: > Also, I think we should align the documentation of this flag with its counter part. The closer they are, the easier to explain things and less maintenance for us.

Unfortunately, `--only` is fundamentally different from `--skip` because it doesn't force all the rules in a group to be executed. Do you think we should change it?

[the remark this answers is not in the document]
Comment (ematipico) [reply crates/biome_cli/src/commands/mod.rs:214]: I would expect both to behave in the same way, logically speaking.

File: crates/biome_analyze/src/lib.rs [hunk 5]
```diff
@@ -843,32 +851,19 @@ impl<'analysis> AnalysisFilter<'analysis> {
...
-    }
-
-    /// It creates a new filter with the set of [enabled rules](RuleFilter) passed as argument
-    pub fn from_enabled_rules(enabled_rules: Option<&'analysis [RuleFilter<'analysis>]>) -> Self {
-        Self {
-            enabled_rules,
-            ..AnalysisFilter::default()
-        }
+    pub fn match_rule<R: Rule>(&self) -> bool {
+        self.enabled_rules.map_or(true, |enabled_rules| {
```
Comment (Conaclos) [inline crates/biome_analyze/src/lib.rs:862]: We don't need to call `match_group` because this check is already handled in `filter.match_rule`.

File: crates/biome_analyze/src/lib.rs [hunk 6]
```diff
@@ -843,32 +851,19 @@ impl<'analysis> AnalysisFilter<'analysis> {
                 enabled_rules.iter().any(|filter| filter.match_group::<G>())
             })
             && self.disabled_rules.map_or(true, |disabled_rules| {
-                !disabled_rules
-                    .iter()
-                    .any(|filter| filter.match_group::<G>())
+                !disabled_rules.iter().any(|filter| {
+                    matches!(filter, RuleFilter::Group(_)) && filter.match_group::<G>()
```
Comment (Conaclos) [inline crates/biome_analyze/src/lib.rs:855]: The implementation of `match_category` was bogus.

File: crates/biome_analyze/src/lib.rs [hunk 7]
```diff
@@ -825,12 +825,20 @@ pub struct AnalysisFilter<'a> {
     /// Only allow rules matching these names to emit signals
     pub enabled_rules: Option<&'a [RuleFilter<'a>]>,
     /// Do not allow rules matching these names to emit signals
-    pub disabled_rules: Option<&'a [RuleFilter<'a>]>,
+    pub disabled_rules: &'a [RuleFilter<'a>],
```
Comment (ematipico) [inline crates/biome_analyze/src/lib.rs:828]: Why did we change the type of `disabled_rules` and not the type of `enabled_rules`? It feels weird that they have two different types. If there's a reason, we should document it.

File: crates/biome_analyze/src/lib.rs [hunk 8]
```diff
@@ -842,33 +850,22 @@ impl<'analysis> AnalysisFilter<'analysis> {
...
+                .any(|filter| matches!(filter, RuleFilter::Group(_)) && filter.match_group::<G>())
     }
 
     /// Return `true` if the rule `R` matches this filter
-    pub fn match_rule<R>(&self) -> bool
-    where
-        R: Rule,
-    {
-        self.match_group::<R::Group>()
+    pub fn match_rule<R: Rule>(&self) -> bool {
```
Comment (ematipico) [inline crates/biome_analyze/src/lib.rs:860]: nit: I am not sure the removal of `where` was necessary...

File: xtask/codegen/src/generate_configuration.rs [hunk 9]
```diff
@@ -141,7 +141,7 @@ pub(crate) fn generate_rules_configuration(mode: Mode) -> Result<()> {
         use biome_json_analyze::options::*;
         use biome_css_analyze::options::*;
         use biome_rowan::TextRange;
-        use indexmap::IndexSet;
+        use rustc_hash::FxHashSet;
```
Comment (ematipico) [inline xtask/codegen/src/generate_configuration.rs:144]: I noticed you're slowly removing `IndexSet` from the code base. Maybe we should do that in another PR; this feels more like a refactor, and unrelated to the PR. But it's fine for now.

**Review by ematipico — APPROVED** [review]
Great work @Conaclos! This is going to be a great feature 🔥

File: crates/biome_analyze/src/lib.rs (hunk 7 above)
Comment (Conaclos) [reply crates/biome_analyze/src/lib.rs:828]: Because it is are fundamentally different from each other. If `disabled_rules` is empty, this means that we disable no rules. Conversely, if  `enabled_rules` is empty, this means that no rule are enabled. In this later case, we need an `Option` to signal the fact that all rules are enabled.

File: crates/biome_analyze/src/lib.rs (hunk 8 above)
Comment (Conaclos) [reply crates/biome_analyze/src/lib.rs:860]: This makes the code cleaner and more consistent with `match_group` (`match_group` doesn't use a `where` clause).

File: crates/biome_analyze/src/lib.rs (hunk 7 above)
Comment (ematipico) [reply crates/biome_analyze/src/lib.rs:828]: Thank you for the explanation. I suppose we should change the docs to explain this part.
